src/stcal/ramp_fitting/likely_fit.py

```diff
@@ -0,0 +1,1335 @@
+
+from . import ramp_fit_class, utils
+from .likely_algo_classes import IntegInfo, Ramp_Result, Covar
+
+
+DELIM = "=" * 80
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+BUFSIZE = 1024 * 300000  # 300Mb cache size for data section
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+def likely_ramp_fit(
+    ramp_data, buffsize, save_opt, readnoise_2d, gain_2d, weighting, max_cores
+):
+    """
+    Setup the inputs to ols_ramp_fit with and without multiprocessing. The
+    inputs will be sliced into the number of cores that are being used for
+    multiprocessing. Because the data models cannot be pickled, only numpy
+    arrays are passed and returned as parameters to ols_ramp_fit.
+
+    Parameters
+    ----------
+    ramp_data : RampData
+        Input data necessary for computing ramp fitting.
+
+    buffsize : int
+        size of data section (buffer) in bytes (not used)
+
+    save_opt : bool
+       calculate optional fitting results
+
+    readnoise_2d : ndarray
+        readnoise for all pixels
+
+    gain_2d : ndarray
+        gain for all pixels
+
+    algorithm : str
+        'OLS' specifies that ordinary least squares should be used;
+        'GLS' specifies that generalized least squares should be used.
+
+    weighting : str
+        'optimal' specifies that optimal weighting should be used;
+         currently the only weighting supported.
+
+    max_cores : str
+        Number of cores to use for multiprocessing. If set to 'none' (the default),
+        then no multiprocessing will be done. The other allowable values are 'quarter',
+        'half', and 'all'. This is the fraction of cores to use for multi-proc. The
+        total number of cores includes the SMT cores (Hyper Threading for Intel).
+
+    Returns
+    -------
+    image_info : tuple
+        The tuple of computed ramp fitting arrays.
+
+    integ_info : tuple
+        The tuple of computed integration fitting arrays.
+
+    opt_info : tuple
+        The tuple of computed optional results arrays for fitting.
+    """
+    image_info, integ_info, opt_info = None, None, None
+
+    nints, ngroups, nrows, ncols = ramp_data.data.shape
+
+    if ngroups < 2:
+        raise ValueError("Likelihood fit requires at least 2 groups.")
+
+    readtimes = get_readtimes(ramp_data)
+
+    covar = Covar(readtimes, pedestal=False)  # XXX Choice of pedestal not given
+    integ_class = IntegInfo(nints, nrows, ncols)
+
+    for integ in range(nints):
+        data = ramp_data.data[integ, :, :, :]
+        gdq = ramp_data.groupdq[integ, :, :, :].copy()
+        pdq = ramp_data.pixeldq[:, :].copy()
+
+        # Eqn (5)
+        diff = (data[1:] - data[:-1]) / covar.delta_t[:, np.newaxis, np.newaxis]
+        alldiffs2use = np.ones(diff.shape, np.uint8)  # XXX May not be necessary
+
+        for row in range(nrows):
+            d2use = determine_diffs2use(ramp_data, integ, row, diff)
+            d2use_copy = d2use.copy()  # Use to flag jumps
```

Comment:
I suggest using the parameter `nsig` from jump detection in the following way:
```suggestion
            threshold_oneomit = nsig**2
            pval = scipy.special.erfc(nsig/2**0.5)
            threshold_twoomit = scipy.stats.chi2.isf(pval, 2)
            if np.isinf(threshold_twoomit):
                threshold_twoomit = threshold_oneomit + 10
```
and then passing `threshold_oneomit` and `threshold_twoomit` in the following line for `mask_jumps`.
This should be very similar to the current pipeline's criteria and let the user tune the knob in a familiar way.  The last line avoids problems if someone chooses `nsig` to be so high that the p value underflows (which will happen if `nsig > 37`).  This is implausible, but I don't think we want to throw an error if the user does something a bit silly like that.  The suggested `if` statement will adopt sensible behavior.